### PR TITLE
Updating IHaskell-widgets

### PR DESCRIPTION
**Pull request by daviddavo**
This pull request aims to solve the issue #870, updating the new widget protocol and implementing any new widgets

Things to do:
- [ ] Update sample notebooks
- [x] Implement the Link widgets
- [x] Implement jslink and jsdlink as helper functions
- [ ] Implement the Controller widget
- [x] Revise and improve documentation
- [x] Fix the OutputWidget

### CODE REVIEW

File: ihaskell-display/ihaskell-widgets/ihaskell-widgets.cabal [hunk 1]
```diff
@@ -110,10 +138,7 @@ library
 
 					 -- The singletons package version is locked to the compiler
 					 -- so let cabal choose the right one.
-                     , singletons -any
-
-  if impl (ghc >= 9.0)
-    build-depends:     singletons-base -any
+                     , singletons < 3.0.0
```
Comment (vaibhavsagar) [inline ihaskell-display/ihaskell-widgets/ihaskell-widgets.cabal:141]: I think this section didn't rebase correctly, can you add what we have in `master` so it works on GHC 9.0?

File: ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Box/HBox.hs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+{-# OPTIONS_GHC -fno-warn-orphans  #-}
+
+module IHaskell.Display.Widgets.Box.HBox
+  ( -- * The HBox widget
+    HBox
+    -- * Constructor
+  , mkHBox
+  ) where
+
+-- To keep `cabal repl` happy when running from the ihaskell repo
```
Comment (vaibhavsagar) [inline ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Box/HBox.hs:15]: Can we remove this?

Comment (vaibhavsagar) [reply ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Box/HBox.hs:15]: Hmm, I see that this same line is present in the codebase prior to this change. I guess it doesn't really do any harm, It's up to you if you want to leave it in or remove it everywhere else too.

Comment (daviddavo) [reply ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Box/HBox.hs:15]: Should I remove also the `mport Prelude`? Or only the comment?

Comment (vaibhavsagar) [reply ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Box/HBox.hs:15]: The `import Prelude` as well, but given how it's all over the `ihaskell-widgets` code it doesn't particularly matter.

**Comment by daviddavo** [comment]
There is a problem with capturing the output using the output widget. The output widget is able to capture the output if you set the `MsgID` attribute to the parent message id of messages to capture. This is thanks to the frontend, so you don't have problems with capturing stdout/stderr. So we could do function like:

```haskell
captureOutput :: OutputWidget -> IO a -> IO ()
captureOutput o f = do
  setField o MsgID (getSomehowTheMsgID)
  f
  setField o MsgID ""
  return ()
```

The problem is, how could we get the id of the `execution_request`?

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -1,5 +1,7 @@
 name: "Docker"
 on:
+  workflow_dispatch:
```
Comment (vaibhavsagar) [inline .github/workflows/docker.yml:3]: Why are you updating the Docker workflow?

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -23,7 +25,7 @@ jobs:
     - uses: actions/checkout@01aecccf739ca6ff86c0539fbc67a7a5007bbc81
     - uses: elgohr/Publish-Docker-Github-Action@f7aca2fea76a5218f3c76cd5933c3ba1d8008774
       with:
-        name: gibiansky/ihaskell
+        name: daviddavo/ihaskell
```
Comment (vaibhavsagar) [inline .github/workflows/docker.yml:28]: I don't think this belongs in this PR, I would suggest making these changes on a separate branch.

File: Widget List.ipynb [hunk 5]
```diff
@@ -0,0 +1,1495 @@
+{
```
Comment (vaibhavsagar) [inline Widget List.ipynb:1]: Please add any extra notebooks to the `notebooks` directory.

File: src/IHaskell/Types.hs [hunk 6]
```diff
@@ -277,6 +295,12 @@ dupHeader hdr messageType = do
   uuid <- liftIO random
   return hdr { mhMessageId = uuid, mhMsgType = messageType }
 
+-- | Modyfies a header and appends the version of the Widget Messaging Protocol as metadata
```
Comment (vaibhavsagar) [inline src/IHaskell/Types.hs:298]:
```suggestion
-- | Modifies a header and appends the version of the Widget Messaging Protocol as metadata
```

File: src/IHaskell/Types.hs [hunk 7]
```diff
@@ -41,8 +43,10 @@ module IHaskell.Types (
 
 import           IHaskellPrelude
 
-import           Data.Aeson (ToJSON (..), Value, (.=), object)
+import qualified Data.HashMap.Strict as HashMap
```
Comment (vaibhavsagar) [inline src/IHaskell/Types.hs:46]: I don't think we need these additional imports if we use `object`.

File: Widget List.ipynb (hunk 5 above)
Comment (daviddavo) [reply Widget List.ipynb:1]: This file was supposed to be ignored as is from IPython...

The widgets notebooks are on ihaskell-display/Examples

File: src/IHaskell/Types.hs [hunk 8]
```diff
@@ -277,6 +295,12 @@ dupHeader hdr messageType = do
   uuid <- liftIO random
   return hdr { mhMessageId = uuid, mhMsgType = messageType }
 
+-- | Modyfies a header and appends the version of the Widget Messaging Protocol as metadata
+setVersion :: MessageHeader  -- ^ The header to modify
+           -> String         -- ^ The version to set
+           -> MessageHeader  -- ^ The modified header
+setVersion hdr v = hdr { mhMetadata = Metadata (HashMap.fromList [("version", String $ pack v)]) }
```
[the remark this answers is not in the document]
Comment (daviddavo) [reply src/IHaskell/Types.hs:302]: The problem is that `Metadata` is `newtype Metadata = Metadata Object`, but the [`object`](https://hackage.haskell.org/package/aeson-1.5.6.0/docs/Data-Aeson.html#v:object) function returns a `Value`.

I saw that they use through the docs the `fromList` function to create an `Object`.

Another alternative could be extracting the `Object` with a lambda expression like `(\(Object o) -> o) $ object ["version .= pack v]`

[the remark this answers is not in the document]
Comment (vaibhavsagar) [reply src/IHaskell/Types.hs:302]: Oh, good point, I didn't realise that. The way you currently do it is fine.

[the remark this answers is not in the document]
Comment (daviddavo) [reply src/IHaskell/Types.hs:302]: I'll add a comment in case we forget about this later

File: ihaskell-display/ihaskell-widgets/README.md [hunk 9]
```diff
@@ -5,3 +5,81 @@ IHaskell. The frontend (javascript) is provided by the jupyter/ipython notebook
...
+- [ ] Overload setField so it can be used with `Maybes` or other wrapper types without having to put `Just` every time.
+- [ ] Add some "utils" work:
+    - [ ] Create media widget from file
+    - [ ] Get the selected label from a selection value
+
+## How to...
+This is a mini-guide for developers that want to update to a more recent widgets specification, but without
+dwelling into the deeps of the project
+
+### Add a new attribute
```
Comment (jamesdbrock) [inline ihaskell-display/ihaskell-widgets/README.md:33]: These instructions are really nice.

File: src/IHaskell/Types.hs (hunk 7 above)
Comment (vaibhavsagar) [reply src/IHaskell/Types.hs:46]: Turns out we can't use `object`.

File: src/IHaskell/Eval/Widgets.hs [hunk 10]
```diff
@@ -18,14 +18,27 @@ import           Control.Concurrent.STM (atomically)
...
+
+import           Data.Foldable (foldl)
 import           System.IO.Unsafe (unsafePerformIO)
 
 import           IHaskell.Display
 import           IHaskell.Eval.Util (unfoldM)
 import           IHaskell.IPython.Types (showMessageType)
 import           IHaskell.Types
 
+#if MIN_VERSION_base(4,11,0)
```
Comment (vaibhavsagar) [inline src/IHaskell/Eval/Widgets.hs:34]: Nit: I'd prefer if we removed this and used plain `fmap` or `<$>` instead.

**Review by vaibhavsagar — APPROVED** [review]
Looks good to me!

File: src/IHaskell/Eval/Widgets.hs (hunk 10 above)
Comment (vaibhavsagar) [reply src/IHaskell/Eval/Widgets.hs]: Resolved!

**Review by jamesdbrock — APPROVED** [review]

**Pull request merged**